mm2src/crypto/src/global_hd_ctx.rs

```diff
@@ -0,0 +1,105 @@
+use crate::privkey::{bip39_priv_key_from_seed, key_pair_from_secret, PrivKeyError};
+use crate::{mm2_internal_der_path, Bip32DerPathOps, Bip32Error, Bip44PathToCoin, CryptoInitError, CryptoInitResult};
+use bip32::{ChildNumber, ExtendedPrivateKey};
+use keys::{KeyPair, Private, Public as PublicKey, Secret as Secp256k1Secret};
+use mm2_err_handle::prelude::*;
+use std::convert::TryInto;
+use std::num::TryFromIntError;
+use std::ops::Deref;
+use std::sync::Arc;
+
+const HARDENED: bool = true;
+const NON_HARDENED: bool = false;
+
+#[derive(Clone)]
+pub struct GlobalHDAccountArc(Arc<GlobalHDAccountCtx>);
+
+impl Deref for GlobalHDAccountArc {
+    type Target = GlobalHDAccountCtx;
+
+    fn deref(&self) -> &Self::Target { &self.0 }
+}
+
+pub struct GlobalHDAccountCtx {
+    bip39_priv_key: ExtendedPrivateKey<secp256k1::SecretKey>,
+    /// secp256k1 key pair derived at `mm2_internal_der_path`.
+    /// It's used for mm2 internal purposes such as signing P2P messages.
+    mm2_internal_key_pair: KeyPair,
```

Comment:
But if the `KeyPairPolicy` is `Iguana`, then `mm2_internal_key_pair` ceases to be "internal", and can be used to enable coins, etc.
On the other hand, `IguanaCtx::secp256k1_key_pair` could very well be the internal key-pair.
Summing up:
`CryptoCtx::mm2_internal_key_pair` shouldn't be called internal because it's not only "internal".
I think it's better to keep the key-pair within `IguanaCtx`.

Also there can be logical problems in the functions like [PrivKeyBuildPolicy::detect_priv_key_policy](https://github.com/KomodoPlatform/atomicDEX-API/blob/mm2-hd-account-secp-secret/mm2src/coins/lp_coins.rs#L2242):
```rust
impl PrivKeyBuildPolicy {
    pub fn detect_priv_key_policy(
        crypto_ctx: &CryptoCtx,
        coin_conf: &Json,
    ) -> MmResult<Self, DetectPrivKeyPolicyError> {
        let secret = match crypto_ctx.key_pair_policy() {
            // THIS: If the build policy is `Iguana`, then we need to use `mm2_internal_privkey_bytes`.
            KeyPairPolicy::Iguana(_) => crypto_ctx.mm2_internal_privkey_bytes(),
            KeyPairPolicy::GlobalHDAccount(hd_ctx) => todo!(),
        };
        Ok(PrivKeyBuildPolicy::Secp256k1Secret(secret))
    }
}
```

Reply:
> Also there can be logical problems in the functions 

Is it really a problem, if it will be encapsulated in the single method? Is it possible to remove iguana Arc at all, so `KeyPairPolicy` will be as follows?
```rust
pub enum KeyPairPolicy {
    Iguana,
    GlobalHDAccount(GlobalHDAccountArc),
}
```
With `detect_priv_key_policy` moved to `CryptoCtx` and maybe also renamed to `get_priv_key_for_coin`?
```rust
impl CryptoCtx {
    pub fn get_priv_key_for_coin(
        &self,
        coin_conf: &Json,
    ) -> MmResult<Self, DetectPrivKeyPolicyError> {
        let secret = match self.key_pair_policy() {
            // THIS: If the build policy is `Iguana`, then we need to use `mm2_internal_privkey_bytes`.
            KeyPairPolicy::Iguana => self.mm2_internal_privkey_bytes(),
            KeyPairPolicy::GlobalHDAccount(hd_ctx) => todo!(),
        };
        Ok(PrivKeyBuildPolicy::Secp256k1Secret(secret))
    }
}
```

Reply:
This definitely makes sense! The only one thing I'd change is to take `derivation_path: Option<&str>` instead of the whole `coin_conf: &Json`.
Will do it!

Reply:
@sergeyboyko0791 Is there a reason to still keep `mm2_internal_key_pair` in `GlobalHDAccountCtx`?